applications/FluidDynamicsApplication/custom_elements/data_containers/derivatives.h

```diff
@@ -0,0 +1,172 @@
+    inline typename std::tuple_element<ElementDataIndex, TCombinedElementDataContainer>::type& GetElementDataContainer(TCombinedElementDataContainer& rCombinedDataContainer) const
+    {
+        static_assert(
+            ElementDataIndex < std::tuple_size<TCombinedElementDataContainer>::value,
+            "Required Element data container index is more than the available element data containers.");
+        return std::get<ElementDataIndex>(rCombinedDataContainer);
+    }
+
+    template<IndexType TAssemblyRowBlockSize, IndexType TAssemblyColumnBlockSize = TAssemblyRowBlockSize>
+    inline void AssembleSubVectorToMatrix(Matrix& rOutput, const IndexType NodeIndex) const
+    {
+        KRATOS_TRY
+
+        KRATOS_DEBUG_ERROR_IF(rOutput.size1() != TAssemblyRowBlockSize * NumNodes)
+            << "rOuput.size1() does not have the required size. [ rOutput.size1() = "
+            << rOutput.size1() << ", required_size = TAssemblyRowBlockSize * NumNodes = "
+            << TAssemblyRowBlockSize * NumNodes << " ].\n";
+
+        KRATOS_DEBUG_ERROR_IF(rOutput.size2() != TAssemblyColumnBlockSize * NumNodes)
+            << "rOuput.size2() does not have the required size. [ rOutput.size2() = "
+            << rOutput.size2() << ", required_size = TAssemblyColumnBlockSize * NumNodes = "
+            << TAssemblyColumnBlockSize * NumNodes << " ].\n";
+
+        for (IndexType i = 0; i < NumNodes; ++i) {
+            for (IndexType j = 0; j < BlockSize; ++j) {
+                rOutput(TAssemblyRowBlockSize * NodeIndex + RowStartingIndex, i * TAssemblyColumnBlockSize + j + ColumnStartingIndex) += mSubVector[i * BlockSize + j];
+            }
+        }
+
+        KRATOS_CATCH("");
+    }
+
+    template<IndexType TAssemblyRowBlockSize>
+    inline void AssembleSubVectorToVector(Vector& rOutput) const
+    {
+        KRATOS_TRY
+
+        static_assert(ColumnStartingIndex == 0);
+
+        KRATOS_DEBUG_ERROR_IF(rOutput.size() != TAssemblyRowBlockSize * NumNodes)
+            << "rOuput.size() does not have the required size. [ rOutput.size() = "
+            << rOutput.size() << ", required_size = TAssemblyRowBlockSize * NumNodes = "
+            << TAssemblyRowBlockSize * NumNodes << " ].\n";
+
+        for (IndexType i = 0; i < NumNodes; ++i) {
+            for (IndexType j = 0; j < BlockSize; ++j) {
+                rOutput[i * TAssemblyRowBlockSize + j] += mSubVector[i * BlockSize + j];
+            }
+        }
+
+        KRATOS_CATCH("");
+    }
+
+    ///@}
+private:
+    ///@name Private members
+    ///@{
+
+    BoundedVector<double, ResidualSize> mSubVector;
+
+    ///@}
+};
+
+template <
+    class TCombinedElementDataContainer,
+    class TCombinedCalculationContainers
+>
+class CalculationDataContainers
+{
+public:
+    ///@name Derivative type definitions
+    ///@{
+
+    using CombinedElementDataContainerType = TCombinedElementDataContainer;
+
+    using CombinedCalculationContainersType = TCombinedCalculationContainers;
```

Comment:
Correct me if I'm wrong, but you need this class to interface to the `CombinedElementDataContainerType` and `CombinedCalculationContainersType` right? If this is the case, I think it would be easier to define these types in the classes you're using them.

Reply:
There are no classes of these types, these are used like the following:

```c++
    using Residual = CalculationDataContainers<
                        std::tuple<
                            Data>,
                        std::tuple<
                            SubAssembly<ResidualsContributions, ElementDataContainerIndex, 0, ResidualColumnOffset>>
                        >;
```

I did not want to create classes so to increase the redundant data.

Reply:
OK. These are only type definitions to be used later on right? In other words, there are no instances of these.

Reply:
Yes... These can be treated as `c++` `traits`. I have changed the name accordingly.